### PR TITLE
Feat: add `se_atten_v2`to PyTorch and DP

### PR DESCRIPTION
**Pull request by Chengqian-Zhang**
Solve #3831 and #3139
- add `se_atten_v2` to PyTorch and DP
- add document equation for `se_attn_v2`

## Summary by CodeRabbit

- **New Features**
  - Introduced a new descriptor class with enhanced configuration options and methods for serialization and deserialization.
  - Added new configurable parameters to the descriptor setup for improved flexibility.

- **Documentation**
  - Updated function documentation to reflect new arguments and usage instructions.

- **Bug Fixes**
  - Refined serialization logic to handle new parameters and class types more accurately.
  - Improved error messages for better clarity during serialization processes.

### CODE REVIEW

**Comment by Chengqian-Zhang** [comment]
@iProzd  thinks there is no need to redirect `dpa1` to `se_atten_v2`. `dpa1` is equivalent to `se_atten` from original definition, while `se_atten_v2` is a special case of `dpa1` to use strip structure. So I'm not redirecting it for now.

File: deepmd/dpmodel/descriptor/dpa1.py [hunk 1]
```diff
@@ -871,6 +871,154 @@ def call(
...
+            concat_output_tebd=concat_output_tebd,
+            spin=spin,
+            stripped_type_embedding=stripped_type_embedding,
+            use_econf_tebd=use_econf_tebd,
+            type_map=type_map,
+            # consistent with argcheck, not used though
+            seed=seed,
+        )
+
+    def serialize(self) -> dict:
```
Comment (njzjz) [inline deepmd/dpmodel/descriptor/dpa1.py:947]: At this time, TF `DescrptSeAttenV2` doesn't have the `serialize` method.

We should add the same method.

**Comment by njzjz** [comment]
Do you want to update the documentation in #3828 or in this PR??

**Comment by Chengqian-Zhang** [comment]
> Do you want to update the documentation in #3828 or in this PR??

I think it is better to update the documentation in this PR.

**Review by wanghan-iapcm — COMMENTED** [review]
could you please provide consistency UT like the se_atten descriptor?

File: doc/model/train-se-atten.md [hunk 2]
```diff
@@ -126,7 +126,7 @@ We highly recommend using the version 2.0 of the attention-based descriptor `"se
...
 ```
 
-When using PyTorch backend, you must continue to use descriptor `"se_atten"` and specify `tebd_input_mode` as `"strip"` and `smooth_type_embedding` as `"true"`, which achieves the effect of `"se_atten_v2"`. The `tebd_input_mode` can take `"concat"` and `"strip"` as values. When using TensorFlow backend, you need to use descriptor `"se_atten_v2"` and do not need to set `tebd_input_mode` and `smooth_type_embedding` because the default value of `tebd_input_mode` is `"strip"`, and the default value of `smooth_type_embedding` is `"true"` in TensorFlow backend. When `tebd_input_mode` is set to `"strip"`, the embedding matrix $\mathcal{G}^i$ is constructed as:
+You can either continue to use descriptor `"se_atten"` and specify `tebd_input_mode` as `"strip"` and `smooth_type_embedding` as `"true"`, which achieves the effect of `"se_atten_v2"`, or use descriptor `"se_atten_v2"` and do not need to set `tebd_input_mode` and `smooth_type_embedding` because the default value of `tebd_input_mode` is `"strip"` and the default value of `smooth_type_embedding` is `"true"`. When `tebd_input_mode` is set to `"strip"`, the embedding matrix $\mathcal{G}^i$ is constructed as:
```
Comment (njzjz) [inline doc/model/train-se-atten.md:129]: > You can either continue to use descriptor `"se_atten"` and specify `tebd_input_mode` as `"strip"` and `smooth_type_embedding` as `"true"`, which achieves the effect of `"se_atten_v2"`,

This may not be required to tell users.

> the default value of `tebd_input_mode` is `"strip"` and the default value of `smooth_type_embedding` is `"true"`

It may not be "default" considering one can not set the parameter when using `se_atten_v2`.

**Comment by njzjz** [comment]
@iProzd could you take a look at the unit test failure?

File: deepmd/tf/descriptor/se_atten.py [hunk 3]
```diff
@@ -1954,8 +1950,8 @@ def serialize(self, suffix: str = "") -> dict:
         }
         if self.tebd_input_mode in ["strip"]:
             assert (
-                type(self) is DescrptDPA1Compat
-            ), "only DescrptDPA1Compat can serialize when tebd_input_mode=='strip'"
+                type(self) is not DescrptSeAtten
+            ), "only DescrptDPA1Compat ansd DescrptSeAttenV2 can serialize when tebd_input_mode=='strip'"
```
Comment (njzjz) [inline deepmd/tf/descriptor/se_atten.py:1954]:
```suggestion
            ), "only DescrptDPA1Compat and DescrptSeAttenV2 can serialize when tebd_input_mode=='strip'"
```

**Review by wanghan-iapcm — APPROVED** [review]

File: deepmd/utils/argcheck.py [hunk 4]
```diff
@@ -631,14 +631,91 @@ def descrpt_se_atten_args():
...
         Argument(
-            "set_davg_zero", bool, optional=True, default=False, doc=doc_set_davg_zero
+            "stripped_type_embedding",
+            bool,
+            optional=True,
+            default=None,
+            doc=doc_stripped_type_embedding,
+        ),
+        Argument(
+            "set_davg_zero", bool, optional=True, default=True, doc=doc_set_davg_zero
```
Comment (iProzd) [inline deepmd/utils/argcheck.py:672]:
```suggestion
            "set_davg_zero", bool, optional=True, default=False, doc=doc_set_davg_zero
```

Comment (iProzd) [reply deepmd/utils/argcheck.py:672]: please notice this default value

**Review by iProzd — CHANGES_REQUESTED** [review]

**Review by iProzd — APPROVED** [review]

File: deepmd/utils/argcheck.py [hunk 5]
```diff
@@ -631,15 +631,92 @@ def descrpt_se_atten_args():
...
+        "Else, the scaling of attention weights is setting to `temperature`."
+    )
+    doc_normalize = (
+        "Whether to normalize the hidden vectors during attention calculation."
+    )
+    doc_concat_output_tebd = (
+        "Whether to concat type embedding at the output of the descriptor."
+    )
+    doc_stripped_type_embedding = (
+        "(Deprecated, kept only for compatibility.) Whether to strip the type embedding into a separate embedding network. "
```
Comment (njzjz) [inline deepmd/utils/argcheck.py:656]: Why is a deprecated option added?

Comment (njzjz) [reply deepmd/utils/argcheck.py:656]: Btw, I remember `tebd_input_mode` is forced to `strip`. Please correct me if I am wrong. @iProzd

Comment (Chengqian-Zhang) [reply deepmd/utils/argcheck.py]: I have deleted the `doc_stripped_type_embedding`. In `"se_atten_v2"`, `tebd_input_mode` is forced to `strip`. So `tebd_input_mode` and `stripped_type_embedding`(and the latter is deprecated) are both useless here.

**Review by njzjz — APPROVED** [review]
